src/mechanics/integrands/MomentumBalance.h

```diff
@@ -23,26 +23,25 @@ class MomentumBalance
     {
     }
 
-    DofVector<double> Gradient(const CellData& cellData, const CellIpData& cellIpData, double deltaT)
+    DofVector<double> Gradient(const CellIpData& cellIpData, double deltaT)
     {
-        BMatrixStrain B = cellIpData.GetBMatrixStrain(mDofType);
-        NodeValues u = cellData.GetNodeValues(mDofType);
         DofVector<double> gradient;
 
-        NuTo::EngineeringStrain<TDim> strain = B * u;
-        gradient[mDofType] = B.transpose() * mLaw.Stress(strain, deltaT, cellData.GetCellId(), cellIpData.GetIpId());
+        BMatrixStrain B = cellIpData.B(mDofType, Nabla::Strain());
+        gradient[mDofType] =
+                B.transpose() * mLaw.Stress(cellIpData.Apply(mDofType, Nabla::Strain()), deltaT, cellIpData.Ids());
```

Comment:
This approach really requires to memoize all operators (in this case B). What is the advantage compared to the "old" version that (at least from my point of view) easier to understand?

Reply:
I do not fully understand the question. 
Do you question the memoization itself? Then I'd say that memoizing `B` and `N` is the whole point of `CellIpData`. 
Or do you question the naming from `GetBMatrixStrain(dof)` to `B(dof, Nabla::Strain())`. Then I'd say that it
- provides a name for different type of `B`s to be used in the `Apply` function
- removes the `B` responsibility from `CellIpData`
- allows to provide custom `B` implementations from outside libnuto